MDNetworking/MDGameSynchronizer.cs

```diff
@@ -0,0 +1,309 @@
+using Godot;
+using System;
+using System.Collections.Generic;
+
+///<summary>A synchronization class with three primary features
+///<para>It tracks the ping to each client</para><para></para>
+///<para>It tracks what the local OS.GetTicksMsec() is on each connected client relative to the server (server only)</para><para></para>
+///<para>Finally it takes care of making sure all players are fully synchronized on request. 
+/// Usually when a player joins or level is changed</para>
+///</summary>
+public class MDGameSynchronizer : Node
+{
+    private const string LOG_CAT = "LogGameSynchronizer";
+
+    public MDGameInstance GameInstance = null;
+
+    public MDGameSession GameSession = null;
+
+    ///<Summary>Stores the latest ping response times for each player</summary>
+    protected Dictionary<int, Queue<float>> InternalPingList = new Dictionary<int, Queue<float>>();
+
+    ///<Summary>Stores all the estimated GetTicksMsec offsets for all players</summary>
+    protected Dictionary<int, List<int>> InternalTicksList = new Dictionary<int, List<int>>();
+
+    ///<Summary>This contains what we think is the GetTicksMsec offset for each player</summary>
+    protected Dictionary<int, int> PlayerTicksMsecOffset = new Dictionary<int, int>();
+
+    ///<Summary>This contains what we think is the ping for each player</summary>
+    protected Dictionary<int, int> PlayerPing = new Dictionary<int, int>();
+
+
+
+    // Called when the node enters the scene tree for the first time.
+    public override void _Ready()
+    {
+        GameSession = GameInstance.GetGameSession();
+        GameSession.OnPlayerJoinedEvent += OnPlayerJoinedEvent;
+        GameSession.OnPlayerLeftEvent += OnPlayerLeftEvent;
+    }
+
+#region PUBLIC METHODS
+
+    /// <summary> Returns the ping for the given peer or -1 if the peer does not exist.</summary>
+    public int GetPlayerPing(int PeerId)
+    {
+        if (!PlayerPing.ContainsKey(PeerId))
+        {
+            MDLog.Warn(LOG_CAT, "Requested ping for peer [{0}] that doesn't exist in list", PeerId);
+            return -1;
+        }
+        return PlayerPing[PeerId];
+    }
+
+    /// <summary> Returns the estimated OS.GetTicksMsec for the given peer or 0 if the peer does not exist.</summary>
+    public uint GetPlayerTicksMsec(int PeerId)
+    {
+        if (!PlayerTicksMsecOffset.ContainsKey(PeerId))
+        {
+            MDLog.Warn(LOG_CAT, "Requested OS.GetTicksMsec for peer [{0}] that doesn't exist in list", PeerId);
+            return 0;
+        }
+        return Convert.ToUInt32(OS.GetTicksMsec() + PlayerTicksMsecOffset[PeerId]);
+    }
+
+#endregion
+
+#region RPC METHODS
+
+    [Puppet]
+    protected void RequestPing(uint ServerTimeOfRequest)
+    {
+        // Respond
+        RpcId(GameSession.GetNetworkMaster(), nameof(PingResponse), MDStatics.GetPeerId(), ServerTimeOfRequest);
```

Comment:
To save you from grabbing the server id, I have a helper: `Node.MDServerRpc` that you can use instead.